File: benchmarks/tasks/contention/contention.hs
-- Threads contending for shared mutable state: eight of them moving money
-- between sixteen accounts, every transfer reading two accounts and writing
-- two as one indivisible step.
--
-- `TVar`s and `atomically`: the same answer Meadow gives, from the language the
-- idea came from.

import Control.Concurrent (forkIO, getNumCapabilities)
import Control.Concurrent.MVar (newEmptyMVar, putMVar, takeMVar)
import Control.Concurrent.STM
import Control.Monad (forM, forM_, when)
import Data.List (intercalate)

accounts, workers, moves :: Int
accounts = 16
workers = 8
moves = 20000

transfers :: [TVar Int] -> Int -> Int -> IO ()
transfers bank s0 n
  | n == 0 = return ()
  | otherwise = do
      let s = s0 * 48271 `mod` 2147483647
          a = s `mod` accounts
          b = s `div` accounts `mod` accounts
          amount = 1 + s `mod` 10
      when (a /= b) $ atomically $ do
        modifyTVar' (bank !! a) (subtract amount)
        modifyTVar' (bank !! b) (+ amount)
      transfers bank s (n - 1)

main :: IO ()
main = do
  bank <- mapM newTVarIO (replicate accounts 1000)
  boxes <- forM [0 .. workers - 1] $ \w -> do
    box <- newEmptyMVar
    _ <- forkIO (transfers bank (w + 1) moves >> putMVar box ())
    return box
  mapM_ takeMVar boxes
  finals <- mapM readTVarIO bank
  putStrLn (intercalate "," (map show finals))

File: benchmarks/tasks/contention/contention.c
/* Threads contending for shared mutable state: eight of them moving money
 * between sixteen accounts, every transfer reading two accounts and writing
 * two as one indivisible step. One mutex over the whole set. */

#include <stdio.h>
#include <stdint.h>
#include <pthread.h>

#define ACCOUNTS 16
#define WORKERS 8
#define MOVES 20000

static int64_t bank[ACCOUNTS];
static pthread_mutex_t lock = PTHREAD_MUTEX_INITIALIZER;

static void *transfers(void *arg) {
    int64_t s = (int64_t)(intptr_t)arg + 1;
    for (int64_t i = 0; i < MOVES; i++) {
        s = s * 48271 % 2147483647;
        int a = (int)(s % ACCOUNTS);
        int b = (int)(s / ACCOUNTS % ACCOUNTS);
        int64_t amount = 1 + s % 10;
        if (a == b) continue;
        pthread_mutex_lock(&lock);
        bank[a] -= amount;
        bank[b] += amount;
        pthread_mutex_unlock(&lock);
    }
    return NULL;
}

int main(void) {
    for (int i = 0; i < ACCOUNTS; i++) bank[i] = 1000;
    pthread_t ts[WORKERS];
    for (intptr_t w = 0; w < WORKERS; w++) pthread_create(&ts[w], NULL, transfers, (void *)w);
    for (int w = 0; w < WORKERS; w++) pthread_join(ts[w], NULL);
    for (int i = 0; i < ACCOUNTS; i++)
        printf("%lld%s", (long long)bank[i], i == ACCOUNTS - 1 ? "\n" : ",");
    return 0;
}
